docs: Add tedge-write documentation page (#3116)
thin-edge/thin-edge.io: docs/src/references/tedge-write.md

---
title: tedge-write
tags: [Reference, Configuration, CLI]

> description: Description of tedge-write component

Author: fixed 14fe31fc6d

---

> **tedge-write** is a helper thin-edge component used by tedge-agent for privilege elevation.

Reviewer: We have a placeholder for thin-edge.io

```suggestion
**tedge-write** is a helper %%te%% component used by tedge-agent for privilege elevation.
```

Author: fixed 14fe31fc6d

tedge-agent spawns a `tedge-write` process when it needs to write to files that `tedge` user/group

> has no write permissions to (e.g. system files or files owned by other packages), for example when

Author: I'd rather keep first sentence as a short summary and explain more in detail in the second sentence.

writing an updated configuration file as part of [`config_update` operation][1]. `tedge-agent` will
first try to write to a file directly and only retry using `tedge-write` if direct write fails due
to `tedge` user/group not having write permissions to either the file itself or its parent

> directory. Write permission to the parent directory is required to guarantee config updates are

Author: I will follow this style more closely.
Usually I use rewrap extension in Vscode which automatically wraps text to a defined number of columns, but it reflows whole paragraphs, so it tends to merge back two sentences when separated by only a single newline, which is a bit unfortunate.

atomic.

[1]: agent/tedge-configuration-management.md#handling-config-update-commands

`tedge-write` will perform an atomic write, i.e. it will write to a temporary file, set file
ownership and mode, and finally rename the temporary into the final file. If the target file already
exists, its original ownership/mode will be preserved and optionally provided new values will be

> ignored.

Reviewer 2: I would organize the doc a bit differently:

- First, the aim of `tedge-write` and when it's used.
- Second, how to configure `sudo` to grant write access to specific files.
- And only then, tell how `tedge-write` works and what it does.

Author: fixed 6ed9c1c9b7

> For `tedge-write` to work, it needs to be able to write to and set ownership and mode of all files.

Author: fixed 14fe31fc6d

To achieve this we can use `sudo` with a `sudoers` configuration that allows `tedge` user to run
`tedge-write` as root without password if the destination file is inside `/etc`:

```sudoers title="file: /etc/sudoers.d/tedge"
tedge   ALL = (ALL) NOPASSWD: /usr/bin/tedge-write /etc/*
```

:::note

> 

Reviewer 3: @albinsuresh which requirement are you reffering to?

> 

Reviewer 4: Uh-oh, that was not supposed to be a PR comment, but it was a suggestion for additional text to be added in the notes. More like:
```suggestion
The default configuration as shown above limits tedge-write to update only the files under the `/etc` directory.
Update this path as per your requirement.
```

> The sudoers entry will be created by thin-edge.io automatically under `/etc/sudoers.d/tedge`, but in
> non-standard setups the default configuration may need to be changed. It's necessary to provide a
> full path to `tedge-write` in the `sudoers` entry. See [`sudoers(5)`][2] for additional details.

Reviewer 4: ```suggestion
When %%te%% is installed using any one of the standard installation methods, the sudoers entry is automatically created under `/etc/sudoers.d/tedge`,
but in non-standard setups the sudoers configuration may need to be updated manually.
It's necessary to provide a full path to `tedge-write` in the `sudoers` entry. See [`sudoers(5)`][2] for additional details.
```

Author: fixed 14fe31fc6d

:::

[2]: https://www.man7.org/linux/man-pages/man5/sudoers.5.html

> If `sudo.enable` is disabled, `tedge-write` will be started without `sudo`.

Reviewer: I think it's better to show a command line example, or at least mentioning `sudo.enable` is a key of tedge config.
I would suggest one example that I came up with, but you don't have to use the same sentence at all.

```suggestion
To start `tedge-write` without `sudo`, set the tedge config `sudo.enable` to `false`.
```

```sh
sudo tedge config set sudo.enable false
```

Author: fixed 14fe31fc6d

## Command help

```sh title="tedge-write"

> A binary used for writing to files which `tedge` user does not have write permissions for, using sudo

Reviewer 4: Nitpick: This message got updated in this PR.

Author: Indeed manually generating command help is quite a churn and easy to forget to update, it would be good to generate automatically in the future so we could easily see help for all commands in the documentation.

pros:
- command help always in sync
- would be easier to verify that flags and options for commands are consistent between different commands

cons:
- more build scripts for the CI pipeline

Author: fixed 14fe31fc6d

Reviewer 2: >  it would be good to generate it automatically

We already have that: [Command block](http://localhost:3000/thin-edge.io/next/contribute/writing-documentation/markdown-components/#command-block)

Author: Replaced with a call to command block, but I'll need to verify when it's deployed to gh-pages that the output is present as when the command could not be started (e.g. because it wasn't found) the build doesn't fail but the text is left empty.
